File: AnalyticsService/Services/AnalyticsService.cs
using System;
using SpotifyTrendsApp.Common.Models;
using AnalyticsService.Models;

namespace SpotifyTrends.AnalyticsService.Services
{
    public class AnalyticsService
    {
        private readonly TopItemsService _topItemsService;
        public AnalyticsService(TopItemsService topItemsService)
        {
            _topItemsService = topItemsService;

        }

        // Add your service methods below
        public async Task<AnalyticsData> GenerateAnalyticsAsync(string timeRange, string authorization)
        {
            // Call TopItemsService to get top tracks and artists
            var topTracks = await _topItemsService.GetTopTracks(timeRange, authorization);
            var topArtists = await _topItemsService.GetTopArtists(timeRange, authorization);


            return new AnalyticsData
            {
                AlbumMosaicData = await GenerateAlbumMosaicAsync(topTracks),
                TopArtistData = await GenerateTopArtist(topTracks),
                ArtistLeaderboardData = await GenerateArtistLeaderboard(topArtists),
                TrackTimelineData = await GenerateTrackTimeline(topTracks),
                GenreBubbleData = await GenerateGenreBubbleChart(topArtists)
            };

        }

        public Task<List<AlbumMosaic>> GenerateAlbumMosaicAsync(List<Track> topTracks)
        {
            var albumMosaic = topTracks
                .GroupBy(track => track.Album.Name)
                .Select(g => new AlbumMosaic
                {
                    AlbumName = g.Key,
                    ArtistName = string.Join(", ", g.First().Artists.Select(artist => artist.Name)),
                    SpotifyLink = g.First().Album.uri ?? string.Empty,
                    Count = g.Count()
                }).ToList();

            return Task.FromResult(albumMosaic);
        }

        public Task<TopArtist> GenerateTopArtist(List<Track> topTracksop)
        {
        // Find the artist that appears most frequently in the tracks
        if (topTracksop == null || topTracksop.Count == 0)
        {
            return Task.FromResult(new TopArtist
            {
                ArtistName = "No Artist Found",
                SpotifyLink = "",
                Count = 0,
                Tracks = Array.Empty<Track>()
            });
        }

        var artistCounts = new Dictionary<string, (int count, string uri, List<Track> tracks)>();

        foreach (var track in topTracksop)
        {
            foreach (var artist in track.Artists)
            {
                if (artist == null || string.IsNullOrEmpty(artist.Name)) continue;
                if (!artistCounts.ContainsKey(artist.Name))
                {
                    artistCounts[artist.Name] = (1, artist.uri ?? string.Empty, new List<Track> { track });
                }
                else
                {
                    var entry = artistCounts[artist.Name];
                    entry.count++;
                    entry.tracks.Add(track);
                    artistCounts[artist.Name] = entry;
                }
            }
        }

        var topArtistEntry = artistCounts.OrderByDescending(a => a.Value.count).FirstOrDefault();

        var topArtist = new TopArtist
        {
            ArtistName = topArtistEntry.Key ?? "No Artist Found",
            SpotifyLink = topArtistEntry.Value.uri,
            Count = topArtistEntry.Value.count,
            Tracks = topArtistEntry.Value.tracks.ToArray()
        };

        return Task.FromResult(topArtist);
        }

        public Task<List<ArtistLeaderboard>> GenerateArtistLeaderboard(List<Artist> topArtists)
        {
            var artistLeaderboard = topArtists
                .Select((artist, index) => new ArtistLeaderboard
                {
                    ArtistName = artist.Name,
                    SpotifyLink = artist.uri ?? string.Empty,
                    Rank = 100 - (index * 2),
                    FollowerCount = artist.Followers,
                    Popularity = artist.popularity
                })
                .OrderByDescending(leaderboard => leaderboard.Rank)
                .ToList();

            return Task.FromResult(artistLeaderboard);
        }

        public Task<List<TrackTimeline>> GenerateTrackTimeline(List<Track> topTracks)
        {
            var trackTimeline = topTracks
                .Where(track => {
                    var releaseDate = track.Album?.release_date;
                    if (string.IsNullOrWhiteSpace(releaseDate)) return false;
                    if (releaseDate == "0" || releaseDate == "0000") return false;
                    DateTime parsedDate;
                    if (!DateTime.TryParse(releaseDate, out parsedDate)) return false;
                    return parsedDate.Year > 1900; // Only allow plausible years
                })
                .Select(track => new TrackTimeline
                {
                    TrackName = track.Name,
                    ArtistName = string.Join(", ", track.Artists.Select(artist => artist.Name)),
                    ReleaseDate = DateTime.TryParse(track.Album.release_date, out var date) ? date : DateTime.MinValue,
                    SpotifyLink = track.uri ?? string.Empty
                })
                .OrderBy(timeline => timeline.ReleaseDate)
                .ToList();

            return Task.FromResult(trackTimeline);
        }

        public Task<List<GenreBubble>> GenerateGenreBubbleChart(List<Artist> topArtists)
        {
            var genreCounts = new Dictionary<string, int>();

            foreach (var artist in topArtists)
            {
                if (artist.Genres != null)
                {
                    foreach (var genre in artist.Genres)
                    {
                        if (genreCounts.ContainsKey(genre))
                        {
                            genreCounts[genre]++;
                        }
                        else
                        {
                            genreCounts[genre] = 1;
                        }
                    }
                }
            }

            var genreBubbles = genreCounts.Select(g => new GenreBubble
            {
                Genre = g.Key,
                Count = g.Value
            }).OrderByDescending(b => b.Count).ToList();

            return Task.FromResult(genreBubbles);
        }


        public Task<MusicDNA> GenerateMusicDNA(List<Track> topTracks, List<Artist> topArtists)
        {
            // TODO: Implement MusicDNA generation logic
            var musicDNA = new MusicDNA
            {
                TopGenres = topArtists.SelectMany(a => a.Genres ??  [""]).Distinct().Take(5).ToList(),
                AveragePopularity = topTracks.Any() ? topTracks.Average(t => t.popularity) : 0,
                TotalTracks = topTracks.Count,
                TotalArtists = topArtists.Count
            };
            return Task.FromResult(musicDNA);
        }

        public Task<List<FollowVsTop>> GenerateFollowedVsTopArtist(List<Track> topTracks, List<Artist> followedArtists)
        {
            // TODO: Implement FollowedVsTopArtist logic
            var followVsTop = new List<FollowVsTop>();
            
            return Task.FromResult(followVsTop);
        } 
    }
}

File: go-services/internal/analytics/service/analytics_service.go
package service

import (
	"sort"
	"strings"
	"time"

	"github.com/sirupsen/logrus"

	"github.com/conor27nash/ListeningTrendsApp/go-services/pkg/models"
)

// AnalyticsService generates analytics from user's top items
type AnalyticsService struct {
	topItemsClient *TopItemsClient
	logger         *logrus.Logger
}

// NewAnalyticsService creates a new AnalyticsService instance
func NewAnalyticsService(topItemsClient *TopItemsClient, logger *logrus.Logger) *AnalyticsService {
	return &AnalyticsService{
		topItemsClient: topItemsClient,
		logger:         logger,
	}
}

// GenerateAnalytics generates all analytics for a given time range
func (s *AnalyticsService) GenerateAnalytics(timeRange, auth string) (*models.AnalyticsData, error) {
	// Fetch data from TopItems service
	tracks, err := s.topItemsClient.GetTopTracks(timeRange, auth)
	if err != nil {
		return nil, err
	}

	artists, err := s.topItemsClient.GetTopArtists(timeRange, auth)
	if err != nil {
		return nil, err
	}

	s.logger.Infof("Generating analytics for %d tracks and %d artists", len(tracks), len(artists))

	// Generate all analytics
	return &models.AnalyticsData{
		AlbumMosaicData:       s.generateAlbumMosaic(tracks),
		TopArtistData:         s.generateTopArtist(tracks),
		ArtistLeaderboardData: s.generateArtistLeaderboard(artists),
		TrackTimelineData:     s.generateTrackTimeline(tracks),
		GenreBubbleData:       s.generateGenreBubbleChart(artists),
	}, nil
}

// generateAlbumMosaic creates album frequency data from tracks
func (s *AnalyticsService) generateAlbumMosaic(tracks []models.Track) []models.AlbumMosaic {
	albumMap := make(map[string]*models.AlbumMosaic)

	for _, track := range tracks {
		albumName := track.Album.Name
		if albumName == "" {
			continue
		}

		if _, exists := albumMap[albumName]; !exists {
			// Create artist names string
			artistNames := make([]string, len(track.Artists))
			for i, artist := range track.Artists {
				artistNames[i] = artist.Name
			}

			albumMap[albumName] = &models.AlbumMosaic{
				AlbumName:   albumName,
				ArtistName:  strings.Join(artistNames, ", "),
				SpotifyLink: track.Album.URI,
				Count:       1,
			}
		} else {
			albumMap[albumName].Count++
		}
	}

	// Convert map to slice
	result := make([]models.AlbumMosaic, 0, len(albumMap))
	for _, mosaic := range albumMap {
		result = append(result, *mosaic)
	}

	return result
}

// generateTopArtist finds the most frequent artist in top tracks
func (s *AnalyticsService) generateTopArtist(tracks []models.Track) models.TopArtist {
	if len(tracks) == 0 {
		return models.TopArtist{
			ArtistName: "No Artist Found",
			Count:      0,
			Tracks:     []models.Track{},
		}
	}

	// Count artist occurrences
	type artistCount struct {
		count  int
		uri    string
		tracks []models.Track
	}

	artistCounts := make(map[string]*artistCount)

	for _, track := range tracks {
		for _, artist := range track.Artists {
			if artist.Name == "" {
				continue
			}

			if _, exists := artistCounts[artist.Name]; !exists {
				artistCounts[artist.Name] = &artistCount{
					count:  1,
					uri:    artist.URI,
					tracks: []models.Track{track},
				}
			} else {
				artistCounts[artist.Name].count++
				artistCounts[artist.Name].tracks = append(artistCounts[artist.Name].tracks, track)
			}
		}
	}

	// Find the artist with the most tracks
	var topArtistName string
	var topCount int
	var topURI string
	var topTracks []models.Track

	for name, data := range artistCounts {
		if data.count > topCount {
			topCount = data.count
			topArtistName = name
			topURI = data.uri
			topTracks = data.tracks
		}
	}

	if topArtistName == "" {
		topArtistName = "No Artist Found"
	}

	return models.TopArtist{
		ArtistName:  topArtistName,
		SpotifyLink: topURI,
		Count:       topCount,
		Tracks:      topTracks,
	}
}

// generateArtistLeaderboard creates ranked artist data
func (s *AnalyticsService) generateArtistLeaderboard(artists []models.Artist) []models.ArtistLeaderboard {
	leaderboard := make([]models.ArtistLeaderboard, len(artists))

	for i, artist := range artists {
		leaderboard[i] = models.ArtistLeaderboard{
			ArtistName:    artist.Name,
			SpotifyLink:   artist.URI,
			Rank:          100 - (i * 2), // Custom ranking: 100, 98, 96, ...
			FollowerCount: artist.Followers,
			Popularity:    artist.Popularity,
		}
	}

	// Sort by rank descending
	sort.Slice(leaderboard, func(i, j int) bool {
		return leaderboard[i].Rank > leaderboard[j].Rank
	})

	return leaderboard
}

// generateTrackTimeline creates chronological track data
func (s *AnalyticsService) generateTrackTimeline(tracks []models.Track) []models.TrackTimeline {
	timeline := []models.TrackTimeline{}

	for _, track := range tracks {
		releaseDate := track.Album.ReleaseDate

		// Validate release date
		if releaseDate == "" || releaseDate == "0" || releaseDate == "0000" {
			continue
		}

		// Try parsing date (Spotify uses YYYY-MM-DD or YYYY)
		var parsedDate time.Time
		var err error

		// Try full date format first
		parsedDate, err = time.Parse("2006-01-02", releaseDate)
		if err != nil {
			// Try year-only format
			parsedDate, err = time.Parse("2006", releaseDate)
			if err != nil {
				// Skip if we can't parse
				continue
			}
		}

		// Filter out old/invalid dates
		if parsedDate.Year() <= 1900 {
			continue
		}

		// Create artist names string
		artistNames := make([]string, len(track.Artists))
		for i, artist := range track.Artists {
			artistNames[i] = artist.Name
		}

		timeline = append(timeline, models.TrackTimeline{
			TrackName:   track.Name,
			ArtistName:  strings.Join(artistNames, ", "),
			ReleaseDate: parsedDate,
			SpotifyLink: track.URI,
		})
	}

	// Sort by date ascending (oldest first)
	sort.Slice(timeline, func(i, j int) bool {
		return timeline[i].ReleaseDate.Before(timeline[j].ReleaseDate)
	})

	return timeline
}

// generateGenreBubbleChart creates genre frequency data
func (s *AnalyticsService) generateGenreBubbleChart(artists []models.Artist) []models.GenreBubble {
	genreCounts := make(map[string]int)

	for _, artist := range artists {
		for _, genre := range artist.Genres {
			if genre != "" {
				genreCounts[genre]++
			}
		}
	}

	// Convert to slice
	bubbles := make([]models.GenreBubble, 0, len(genreCounts))
	for genre, count := range genreCounts {
		bubbles = append(bubbles, models.GenreBubble{
			Genre: genre,
			Count: count,
		})
	}

	// Sort by count descending
	sort.Slice(bubbles, func(i, j int) bool {
		return bubbles[i].Count > bubbles[j].Count
	})

	return bubbles
}
